### PR TITLE
SOC-786 Create MediaWiki Client For User Attribute Service

### PR DESCRIPTION
**Pull request by jsutterfield**
This ticket used swagger-codegen to create a php client for communicating with the User Attribute Service (see [here](https://github.com/Wikia/app/blob/dev/lib/Swagger/README.md) for more info on the swagger-codegen). That client was then configured to use the dependency injection library the services team set-up.

Ticket: https://wikia-inc.atlassian.net/browse/SOC-786

### CODE REVIEW

File: includes/User.php [hunk 1]
```diff
@@ -23,6 +23,7 @@
 use Wikia\DependencyInjection\Injector;
 use Wikia\Logger\Loggable;
 use Wikia\Service\User\Preferences\UserPreferences;
+use Wikia\Service\User\Attributes\UserAttributes;
```
Comment (jsutterfield) [inline includes/User.php:26]: Whoa! This is a huge pull request!

Fear not, fair code reviewer. The bulk of this is the auto-generated client code created by swagger-autogen which looks at the Swagger annotations in the User Attribute Service. The code which I added to wire that client into MW starts [here](https://github.com/Wikia/app/pull/7943/files#diff-5f12f40345b8d749abbb52d78fda227eR1) which is much more manageable.

File: lib/Wikia/src/DependencyInjection/InjectorInitializer.php [hunk 2]
```diff
@@ -12,6 +13,7 @@ public static function init(CacheProvider $cacheProvider = null) {
 			(new InjectorBuilder())
 				->withCache($cacheProvider)
 				->addModule(new PreferenceModule())
+				->addModule(new AttributesModule())
```
Comment (jsutterfield) [inline lib/Wikia/src/DependencyInjection/InjectorInitializer.php:16]: If you're not familiar with the dependency injection library we just added (which I assume you're not b/c it was just added, I wasn't), this is the entry point.

When MW lurches into existence, this Attributes Module is called which binds a set of interfaces to concrete classes which implement those interfaces. It also binds the class constant `UserAttributes::DEFAULT_ATTRIBUTES` to `\User::getDefaultOptions();`. All this allows the various dependencies to be mocked easily for testing (which you can see in the unit tests below).

Instantiating the client is done through the DI library [here](https://github.com/Wikia/app/pull/7943/files#diff-897fba1dd6b8edc3238762e35d0b7c71R275) which takes care of resolving all the dependencies, and injecting instances of those dependencies.

File: lib/Wikia/src/Service/User/Attributes/AttributesModule.php [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+use Wikia\DependencyInjection\Module;
+use Wikia\Persistence\User\Attributes\AttributePersistence;
+use Wikia\Persistence\User\Attributes\AttributePersistenceSwagger;
+
+class AttributesModule implements Module {
+	public function configure( InjectorBuilder $builder ) {
+		$builder
+			->bind( AttributeService::class )->toClass( AttributeKeyValueService::class )
+			->bind( AttributePersistence::class )->toClass( AttributePersistenceSwagger::class )
+			->bind( UserAttributes::DEFAULT_ATTRIBUTES )->to( function() {
```
Comment (garthwebb) [inline lib/Wikia/src/Service/User/Attributes/AttributesModule.php:15]: When does this become relevant?  Does the `@Inject({` comment below get treated as an annotation somehow?

**Comment by garthwebb** [comment]
LGTM

File: lib/Wikia/src/Service/ForbiddenException.php [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<?php
+
+namespace Wikia\Service;
+
+class ForbiddenException extends \Exception {
+	const CODE = 403;
+}
```
Comment (kvas-damian) [inline lib/Wikia/src/Service/ForbiddenException.php]: Missing new line at the end of file

File: lib/Wikia/src/Service/User/Attributes/AttributeKeyValueService.php [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+		$this->persistenceAdapter = $persistenceAdapter;
+	}
+
+	public function setAttribute( $userId, $attribute ) {
+		if ( empty( $attribute ) || $userId == 0 ) {
+			return false;
+		}
+
+		try {
+			$profiler_start = $this->startProfile();
```
Comment (kvas-damian) [inline lib/Wikia/src/Service/User/Attributes/AttributeKeyValueService.php]: our PHP guidelines suggest lowerCamelCase for variable names.

File: lib/Wikia/src/Service/User/Attributes/AttributeKeyValueService.php [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+	 *    Wikia\Persistence\User\Attributes\AttributePersistence::class
+	 * })
+	 * @param AttributePersistence $persistenceAdapter
+	 */
+	function __construct( AttributePersistence $persistenceAdapter ) {
+		$this->persistenceAdapter = $persistenceAdapter;
+	}
+
+	public function setAttribute( $userId, $attribute ) {
+		if ( empty( $attribute ) || $userId == 0 ) {
```
Comment (kvas-damian) [inline lib/Wikia/src/Service/User/Attributes/AttributeKeyValueService.php]: What do you think about strict comparison `===`?

Comment (kvas-damian) [inline lib/Wikia/src/Service/User/Attributes/AttributeKeyValueService.php]: ~~Why it's not possible to set empty value for attribute? 
I can imagine setting attribute flags - (`true/false` value). In this case I cannot set false value. Additionally there is no way to delete attribute right now.~~

Comment (kvas-damian) [inline lib/Wikia/src/Service/User/Attributes/AttributeKeyValueService.php]: NVM I misinterpreted code ;)

File: lib/Wikia/src/Service/User/Attributes/AttributeKeyValueService.php [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+			return false;
```
Comment (kvas-damian) [inline lib/Wikia/src/Service/User/Attributes/AttributeKeyValueService.php]: Shouldn't we throw an exception in this case?

File: lib/Wikia/src/Service/User/Attributes/AttributeKeyValueService.php [hunk 8]
```diff
@@ -0,0 +1,74 @@
+<?php
+
+namespace Wikia\Service\User\Attributes;
+
+use Wikia\Logger\Loggable;
+use Wikia\Persistence\User\Attributes\AttributePersistence;
+use Wikia\Util\WikiaProfiler;
+use Wikia\Service\User\Attributes;
+
+class AttributeKeyValueService implements AttributeService {
```
Comment (kvas-damian) [inline lib/Wikia/src/Service/User/Attributes/AttributeKeyValueService.php:10]: Don't we need `delete` method?

File: lib/Wikia/src/Service/User/Attributes/AttributeKeyValueService.php [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+	 * @Inject({
+	 *    Wikia\Persistence\User\Attributes\AttributePersistence::class
+	 * })
+	 * @param AttributePersistence $persistenceAdapter
+	 */
+	function __construct( AttributePersistence $persistenceAdapter ) {
+		$this->persistenceAdapter = $persistenceAdapter;
+	}
+
+	public function setAttribute( $userId, $attribute ) {
```
Comment (kvas-damian) [inline lib/Wikia/src/Service/User/Attributes/AttributeKeyValueService.php]: While we're in `AttributeKeyValueService` class what do you think about making method names more simple, eg. `set`, `get`?

File: lib/Wikia/src/Service/User/Attributes/AttributeKeyValueService.php (hunk 8 above)
Comment (jsutterfield) [inline lib/Wikia/src/Service/User/Attributes/AttributeKeyValueService.php:10]: I added it in the API for the service, but I don't think it's a very common operation. From what I can tell the User class doesn't even expose a way to do it, you can get and update attributes, but not delete.

That being said, it is a functionality we expose in the API so we should probably support it in the client as well. I'll update.

File: lib/Wikia/src/Service/User/Attributes/AttributeKeyValueService.php (hunk 9 above)
Comment (jsutterfield) [inline lib/Wikia/src/Service/User/Attributes/AttributeKeyValueService.php]: Sure, that sounds fine to me.

File: lib/Wikia/src/Service/User/Attributes/AttributesModule.php (hunk 3 above)
Comment (jsutterfield) [inline lib/Wikia/src/Service/User/Attributes/AttributesModule.php:15]: Yea, exactly. The DI library looks at that annotation to determine what to inject. It sees that we want to inject `UserAttributes::DEFAULT_ATTRIBUTES` which is bound to this function, so it injects the result of this function.

File: lib/Wikia/src/Service/ForbiddenException.php (hunk 4 above)
Comment (jsutterfield) [inline lib/Wikia/src/Service/ForbiddenException.php]: Added new line.

File: lib/Wikia/src/Service/User/Attributes/AttributeKeyValueService.php [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+	 * })
+	 * @param AttributePersistence $persistenceAdapter
+	 */
+	function __construct( AttributePersistence $persistenceAdapter ) {
+		$this->persistenceAdapter = $persistenceAdapter;
+	}
+
+	public function set( $userId, $attribute ) {
+		if ( empty( $attribute ) || $userId === 0 ) {
+			return false;
```
Comment (kvas-damian) [inline lib/Wikia/src/Service/User/Attributes/AttributeKeyValueService.php]: There was a question to this line which most likely got missed after code changes ;)
Shouldn't we throw an exception in this case?

Comment (jsutterfield) [inline lib/Wikia/src/Service/User/Attributes/AttributeKeyValueService.php]: Yep, that definitely got missed :) Sure, that works too. A lot of this code was heavily inspired by the User Preferences Service (read, borderline plagiarized), so this was a hold over from that. An exception sounds good there. I'll add it.

**Comment by kvas-damian** [comment]
LGTM

**Pull request merged**